### ISSUE TITLE
Team Shared Connections without Passwords, but still allow users to save passwords

### ISSUE THREAD

**Issue by jadzdotcom** [serge-rider/dbeaver#1083]
I've been investigating using dbeaver with a team of users to try to share a standard set of connections and have a common tool for everyone. 

Admin Managed Connections  (https://github.com/serge-rider/dbeaver/wiki/Admin-Manage-Connections) works well for this if either I want everyone to have the passwords to all connections, or if I don't want to share any passwords and I don't want to allow users to be able to save passwords locally. Sharing the passwords with everyone doesn't make sense in my scenario, nor does forcing users to enter passwords each time they connect as this greatly reduces adoption.

What I would like to do is:
* Share connection definition (i.e. .dbeaver-data-sources.SHARED.xml) without passwords (from source control)
* Allow users to save passwords for connections without modifying the shared connection file (so if they update from source control, they get new/updated connections without affecting their locally saved passwords).

I was hopeful when I saw the recent secure password features of EE, that my use case might be covered off just by turning on this option, but it doesn't appear to work that way.

Am I missing a way to do this with the current features?

**Comment by serge-rider** [comment]
New secured passwords feature uses OS facilities + external passwords storage. Generally this should work for you purpose.
Could you clarify "doesn't appear to work that way" thing?

It won't work if user will change user name (because name is still stored in datasources.xml). Maybe I should move user names into secured storage too..

**Comment by jadzdotcom** [comment]
When I enabled secured passwords (on a Windows computer) it prompted me for the password, and would save it for the session, but when I restart dbeaver it would forget the password and then force me to re-enter it. 

It is possible that there is something odd with the connection definition. I can try to create with a fresh install/connection and see if I get the same behaviour.

**Comment by jadzdotcom** [comment]
I just did some testing on a clean environment. 

If "Save password locally" is checked, then when the user tries to save edit the connection it provides the error "This connection was provided by system administrator. Its settings can not be changed." If the "Save password locally" is unchecked (for the admin managed connection) then it will not save the password and will re-prompt the next time you start DBeaver.

**Comment by serge-rider** [comment]
Secured passwords storage uses master password (it will be asked every time you start DBeaver). On Windows platforms you can use native encryption (configurable in preferences) so current user's password will be used as master password.

Yes, currently editing of "provided" datasources is not supported in 3.8.2. I'll fix this so it will be possible to edit such connections. Also passwords encryption will be improved (user name encryption + write protecton of whole datasources.xml file with password hash).

**Comment by GiantCrocodile** [comment]
@serge-rider as this is the issue why I became aware of this: Why do we use a fixed `PASSWORD_ENCRYPTION_KEY` and not a randomized one in https://github.com/serge-rider/dbeaver/blob/b5b4bf9864e68af07c9fa2c9143487497c53326f/plugins/org.jkiss.dbeaver.core/src/org/jkiss/dbeaver/registry/encode/SimpleStringEncrypter.java#L32 ?

**Comment by serge-rider** [comment]
Because datasources configuration supposed to be portable (you can pass it to other people and it will work). Base passwords encryption isn't a real protection (mainly it was made to avoid plain-text passwords storage).
So if you aware about your saved passwords security - you should use this new feature.

**Comment by GiantCrocodile** [comment]
A fixed key is like a plain-text password: if you use the same key anywhere you can read the password anywhere. It doesn't really make the situation harder for a thief/attacker. I suggest to generate a key with unknown variables (random stuff like time or mouse coordinates or user based input variables) on first start up/use of this feature and let the person who import this portable data source configuration to define the key on import. Thus you have a secure key for every client who uses DBeaver and it is still portable. (Could be that I misunderstand something and thus my suggestion is false/bad.)

**Comment by serge-rider** [comment]
I don't think this make any sense. I don't see real difference with current approach. You'll have to keep this random key somewhere (in some known location) to decrypt passwords. 
This is just one tiny little step for thief but a real headache for users. What's the point?

If you need to protect password on your local machine - use secure storage with master password. This is the only more or less secure way. 
If you need even more security - don't save passwords at all.

**Comment by jadzdotcom** [comment]
While keeping both the username and password in local secured storage will likely address another use case (i.e. not mine), are you making a change that will somehow associate the saved user/password with the connection? 

My thought is that if you can store the user/password in secured storage with the data-source id, or name, then it could be used with an administrator provided connection definition without having to modify the actual .dbeaver-data-sources.ADMIN.xml file. 

I suspect that just moving the username into the local secured storage will solve the blocking issue that I'm seeing in the current 3.8.2 release (If "Save password locally" is checked, then when the user tries to save edit the connection it provides the error "This connection was provided by system administrator. Its settings can not be changed." If the "Save password locally" is unchecked (for the admin managed connection) then it will not save the password and will re-prompt the next time you start DBeaver).

**Comment by serge-rider** [comment]
Yes, the issue with unsavable .dbeaver-data-sources.ADMIN.xml was also resolved.
This was a pretty old malfunction, not related to secure storage.
